src/Microsoft.ML.Data/Data/SchemaDefinition.cs

```diff
@@ -326,21 +326,12 @@ public enum Direction
             Both = Read | Write
         }
 
-        /// <summary>
-        /// Create a schema definition by enumerating all public fields of the given type.
-        /// </summary>
-        /// <param name="userType">The type to base the schema on.</param>
-        /// <param name="direction">Accept fields and properties based on their direction.</param>
-        /// <returns>The generated schema definition.</returns>
-        public static SchemaDefinition Create(Type userType, Direction direction = Direction.Both)
+        public static MemberInfo[] GetMemberInfos(Type userType, Direction direction)
```

Comment:
Does the new methods that you introduced in `SchemaDefinition `need to be `public`? Can they be `internal`?

Since SchemaDefinition is a public class (i.e. it can be accessed by our users), then adding new public methods will also expose them to the users, which might not be desirable: in general I don't see why these need to be accessed by the users (they'd only need to access the Create method, which was already public, to create their own SchemaDefinitions). 

Also, in ML.NET we have a commitment of never making any changes that will break the public API until ML.NET version 2. So, adding these public methods, mean that from now on, we won't be able to delete them, or rename them, or change its parameters... since they'll belong to the public API.

In order to avoid this, we typically use the `internal` access modifier. By using it, these methods won't be publicly accessible to users, and they won't be considered members of our public API.

Reply:
BTW, this doesn't need to be changed in the InternalSchemaDefinition, because there even if the methods are public they belong to the `internal` class `InternalSchemaDefinition`, so they aren't publicly available to users.

---
In reply to: [431450626](https://github.com/dotnet/machinelearning/pull/5114#discussion_r431450626) [](ancestors = 431450626)